src/NuGet.Core/NuGet.Versioning/NuGetVersionFactory.cs

```diff
@@ -87,6 +101,16 @@ public static bool TryParse(string? value, [NotNullWhen(true)] out NuGetVersion?
                             metadata: buildMetadata ?? string.Empty,
                             originalVersion: originalVersion);
 
+                        lock (ParsedNuGetVersionsMapping)
```

Comment:
Do you have measurements on approximately how much contention we'd expect here? Would it be worth considering `ConcurrentDictionary`?

Reply:
This particular call on the lock shouldn't hit much contention, however, the one at the beginning of the method could. Keep in mind that the code executed in that case is very light, just a quick dictionary lookup.

From my understanding, ConcurrentDictionary.Count is pretty bad performance wise. The usage of the lock was simple enough, combined with the Count() concern, that this seemed like the right approach.

Reply:
Sounds good to me :)

Reply:
@jeffkl  I'm not sure there'd be much savings in this instance. We decided to take that approach with the `TaskResultCache` since kicking off the tasks can have arbitrary synchronous code execute before the task yields. The dictionary lookups/adds will be rather fast so there should be minimal contention unless this specific path gets absolutely hammered.

Reply:
Tracking at a per-key level, feels like re-implementing ConcurrentDictionary. If you feel fairly strong, I could just switch to ConcurrentDictionary and manually keep a best effort count manually.

Reply:
My vote would be to keep the current implementation.

Reply:
Additional context I forgot to mention. The reason we went with a lock per key approach in the `TaskResultCache` was because of the guarantees we needed to provide:

1. Only one canonical task can be created per key
2. We want to be flexible enough to handle somewhat arbitrary use
3. Lock contention should be minimized.

A "quirk" of `ConcurrentDictionary` is the value factory is run *outside* of the lock internally when calling `GetOrAdd()`. You're guaranteed that each caller gets the same item back, but the item can be created multiple times. It was implemented this way to avoid holding the lock for longer running factory delegates. In `TaskResultCache`, we use a canonical lock object per key to allow more concurrency without creating multiple tasks for the same key.